### PR TITLE
yiyan prepare dataset v1

### PR DESCRIPTION
**Pull request by yiyanh**
Added `scripts/prepare_breast_path.py`: generate train.pt, validation.pt, test.pt from raw annotated dataset.
By default, `python scripts/prepare_breast_path.py` reads a csv file and generates output files in `data/breast_path/`
So far, the script generates output files of the correct format, but max_seq_length error occurs when running finetuning.

### CODE REVIEW

File: scripts/prepare_breast_path.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+
+    for l in lateralities:
+        for t in types:
+            k+=1
+            comb = f"If {l} is being examined, is there cancer of type {t} found in {l}? Answer 1 if yes, 0 if no.\nAnswer:" 
+            combs[k] = comb
+    
+    prompts = []
+    responses = []
+    questions = [] # question numbering for balancing tr/val
```
Comment (yala) [inline scripts/prepare_breast_path.py:137]: not clear what this means

File: scripts/prepare_breast_path.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+    q1q2 = [question_1, question_2]
+    for i in range(2):
+        p = prefix + row.loc["Report Text"] + q1q2[i]
+        prompts.append(p)
+        responses.append(row.iloc[i+2])
+        questions.append(i+1)
+
+    if row.loc["q3"]!=0: # only creates QA pair for Q3 when there is cancer
```
Comment (yala) [inline scripts/prepare_breast_path.py:145]: might be easier to just always ask the question and have the answer be 0 if there is no cancer, its still correct and makes the schema flatter

File: scripts/prepare_breast_path.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        p = prefix + row.loc["Report Text"] + q1q2[i]
+        prompts.append(p)
+        responses.append(row.iloc[i+2])
+        questions.append(i+1)
+
+    if row.loc["q3"]!=0: # only creates QA pair for Q3 when there is cancer 
+        for k, v in combs.items():
+            p = prefix + row.loc["Report Text"] + v
+            prompts.append(p)
+            questions.append(3)
```
Comment (yala) [inline scripts/prepare_breast_path.py:149]: huh, why 3

File: scripts/prepare_breast_path.py [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+            if isinstance(row.loc["q3"], int) and row.loc["q3"]==k:
+                responses.append(1)
+            elif isinstance(row.loc["q3"], list) and k in row.loc["q3"]:
+                responses.append(1)
+            else:
+                responses.append(0)
```
Comment (yala) [inline scripts/prepare_breast_path.py:155]: this is all hard to read

File: scripts/prepare_breast_path.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    result = []
+    if len(responses)!=n or len(questions)!=n:
+        print("Something wrong! # prompts & # responses do not match.\n")
+    else:
+        for i in range(n):
+            result.append({"id": row.loc["Accession Number"],
+                           "prompt": prompts[i],
+                           "answer": responses[i],
+                           "q": questions[i]})
+        return result
```
Comment (yala) [inline scripts/prepare_breast_path.py:167]: This whole fn is more complicated than it needs to be.
Long term, you want this to be configurable and have like a JSON / YAML template that this code reads and automatically does something reasonable

If the fn is this complex, you can also break it down into parts with comments per part. Always name things semantically and don't append random ints like 3

File: scripts/prepare_breast_path.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    
+
+###Preprocess annotated pathology dataset############################################################
```
Comment (yala) [inline scripts/prepare_breast_path.py:170]: comments should be bellow fn definiton.
See other examples for doc strings in the lightinig code

def fn_name():
"""
    description 
    args: 
          thing: Descrip
"""

Comment (yala) [reply scripts/prepare_breast_path.py:170]: also, using types is nice in args

File: scripts/prepare_breast_path.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+def preprocess(df, split_size, seed):
+    df = df[df["Comment"].isna()] # 24674 --> 24532
```
Comment (yala) [inline scripts/prepare_breast_path.py:172]: these comments don't make sense

File: scripts/prepare_breast_path.py [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+    df = df[df["Report Text"].str.contains('MOLECULAR PATHOLOGY REPORT')==False]
+    for index, row in df.iterrows():
+        # Extract the text 
+        text = row["Report Text"]  
+
+        if row["Exam Description"] == "FINE NEEDLE ASPIRATION (DELIVER SPECIMEN TO PATHOLOGY)":
+            text = extract_text_fn(text)
+        
+        elif row["Exam Description"] == "SURGICAL PATHOLOGY":
+            text = extract_text_sp(text)
```
Comment (yala) [inline scripts/prepare_breast_path.py:185]: could these fn's be named better?

Also can probably do this as cleaner 2 line:
```
descrip_to_fn = { "fine need ..." : extract_text_fine_needle, ..
text = descrip_to_fn[ exam[descrip] ](exam_
```

File: scripts/prepare_breast_path.py [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+        
+        df.at[index, "Report Text"] = text
+
+    # summarize answer to 3 questions
+    df["q1"] = df.apply(answer_q1, axis=1)
+    df["q2"] = df.apply(answer_q2, axis=1)
+    df["q3"] = df.apply(answer_q3, axis=1)
+    df = df[["Accession Number", "Report Text", "q1", "q2", "q3"]]
```
Comment (yala) [inline scripts/prepare_breast_path.py:193]: bad naming 
also, no idea what this means

File: scripts/prepare_breast_path.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+    print("Dataset preprocessed\n")
+    
+    return split(df, split_size=split_size, seed=seed)
+
+def extract_text_fn(text):
```
Comment (yala) [inline scripts/prepare_breast_path.py:198]: add comments to explain what this is supposed to do with a mini demo example

File: scripts/prepare_breast_path.py [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+        match4 = re.search(pattern, result, re.DOTALL)
+        matches.append(match4)
+
+    for m in matches:
+        if m:
+            results.append(m.group(0))
+    
+    return min(results, key=len)
+
+def extract_text_sp(text):
```
Comment (yala) [inline scripts/prepare_breast_path.py:231]: name this better. variable names are cheap, you have auto-compelte so don't abreviate too much

File: scripts/prepare_breast_path.py [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+    result = text
+    pattern = r'(?:(?<=PATHOLOGIC DIAGNOSIS)|(?<=FINAL DIAGNOSIS)|(?<=CYTOLOGIC DIAGNOSIS)|(?<=FINAL PATHOLOGIC DIAGNOSIS))([\s\S]*)'
```
Comment (yala) [inline scripts/prepare_breast_path.py:233]: shared things should be CONSTS at the top of the file

File: scripts/prepare_breast_path.py [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+    
+    for m in matches:
+        if m:
+            results.append(m.group(0))
+    
+    return min(results, key=len)
+
+def answer_q1(row): # returns 0 if no source, an integer if one source, a list if multiple sources
+    answer = []
+    if row.loc["Source_LB"] == 1:
```
Comment (yala) [inline scripts/prepare_breast_path.py:274]: this could be much more consice:

```
keys = [Source LB, Source RB, ...]
for i, key in enumerate(keys):
      answer.append(i) if row.loc[key] == 1 
```

The end could be more concise and cleaner too

File: scripts/prepare_breast_path.py [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+    if row.loc["Source_O"] == 1:
+        answer.append(5)
+    if len(answer) == 0:
+        return 0
+    elif len(answer) == 1:
+        return answer[0]
+    else:
+        return answer
+    
+def answer_q2(row): # returns 1 if there is cancer, 0 if not
```
Comment (yala) [inline scripts/prepare_breast_path.py:291]: bad naming and wrogn doc string format

File: scripts/prepare_breast_path.py [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+    return int(np.nansum(row[["Diagnosis_LB", "Diagnosis_RB", "Diagnosis_LL", "Diagnosis_RL"]]) > 0)
+
+def answer_q3(row): # returns 0 if no cancer, 1-12 for each combo of tissue source x cancer type
+    tissues = ["Type_LB", "Type_RB"]
+    answer = []
+    #types = ["ductal carcinoma in situ (DCIS)", "invasive ducal carcinoma (IDC)", "invasive lobular carcinoma (ILC)", "other invasive", "adenocarcinoma", "other"]
```
Comment (yala) [inline scripts/prepare_breast_path.py:297]: dont have commented code in the PRs

**Review by yala — CHANGES_REQUESTED** [review]
For the revision, please conciosuly make an effort to write your cleanest and best code.
The point of me reviewing the code is to help to push max quality so we can succesffuly go through the project

Shorter / cleaner code is always easier to debug, maintain and to get help on that longer complicated code

File: .gitignore [hunk 16]
```diff
@@ -8,7 +8,6 @@ build
 
 # data
 data
-datasets
```
[the remark this answers is not in the document]
Comment (yiyanh) [reply .gitignore:11]: There isn't such a folder

File: scripts/prepare_breast_path.py (hunk 3 above)
Comment (yiyanh) [reply scripts/prepare_breast_path.py:149]: This was for labeling which question we are asking the model, since I wanted to distinguish the 12 subquestions in question 3 (cancer type) from questions 1 (tissue source) and 2 (cancer presence). I stored this information more for question-specific accuracy calculation, as well as creating training set that have enough data points for all 3 questions. Otherwise I worry question 1/2 may not be selected as often as question 3.

File: .gitignore (hunk 16 above)
[the remark this answers is not in the document]
Comment (yala) [reply .gitignore:11]: it can a folder for other team members though, so in general, it's not good practice to remove stuff from gitignore since other people can name folders different from you
